File: backend/cs/SocialNetApp/Common.MQ/Core/Services/MQSender.cs
using Common.MQ.Core.Model;
using Common.MQ.Core.Model.Interfaces;
using Common.MQ.Core.Model.Types;
using RabbitMQ.Client;
using System.Text;
using System.Text.Json;
using System.Threading.Channels;

namespace Common.MQ.Core.Services
{
    public class MQSender : MQBase, IMQSender
    {
        public void SendNewDialogMessage<T>(string queuePostfix, T data)
        {
            SendDirectMessage($"{MQConstants.DialogWebsockQueueName}_{queuePostfix}", data);
        }

        public void SendUnreadDialogMessageIds(uint userId, bool isIncrement, int[] unreadMsgIds)
        {
            SendDirectMessage(MQConstants.CounterQueueName, 
                new UnreadCountMessage { 
                    MessageType = MQMessageTypes.UpdateUnreadDialogMessages,
                    UserId = userId,
                    IsIncrement = isIncrement,
                    UnreadMessageIds = unreadMsgIds
                });
        }

        public void SendUnreadDialogMessageIdsFailed(uint userId, bool isIncrement, int[] unreadMsgIds)
        {
            SendDirectMessage(MQConstants.CounterDialogQueueName,
                new UnreadCountMessage
                {
                    MessageType = MQMessageTypes.UpdateUnreadDialogMessagesCompensate,
                    UserId = userId,
                    IsIncrement = isIncrement,
                    UnreadMessageIds = unreadMsgIds
                });
        }        

        public void SendPost(uint userId, string post)
        {
            var (connection, channel) = CreateMQ();

            channel.ExchangeDeclare(exchange: MQConstants.PostChannelName, type: ExchangeType.Topic);

            var routingKey = userId.ToString();
            var body = Encoding.UTF8.GetBytes(post);
            channel.BasicPublish(exchange: MQConstants.PostChannelName,
                                 routingKey: routingKey,
                                 basicProperties: null,
                                 body: body);

            channel?.Dispose();
            connection?.Dispose();
        }

        private void SendDirectMessage<T>(string queueName, T data)
        {
            var (connection, channel) = CreateMQ();
            channel.QueueDeclare(queue: queueName,
                durable: false,
                     exclusive: false,
                     autoDelete: false,
                     arguments: null);
            var text = JsonSerializer.Serialize(data);
            var body = Encoding.UTF8.GetBytes(text);
            channel.BasicPublish(exchange: string.Empty,
                                 routingKey: queueName,
                                 basicProperties: null,
                                 body: body);

            channel?.Dispose();
            connection?.Dispose();
        }
    }
}

File: backend/go/socialnetapp/pkg/mq/mqsender.go
package mq

import (
	"context"
	"log"
	"os"
	"strconv"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

type mqSenderImp struct {
}

func NewMqSender() MqSender {
	return &mqSenderImp{}
}

func (s *mqSenderImp) SendPost(_ context.Context, userId uint, post string) error {
	conn, err := amqp.Dial(os.Getenv("RABBITMQ_CONNECTION"))
	if err != nil {
		return err
	}
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil {
		return err
	}
	defer ch.Close()

	err = ch.ExchangeDeclare(
		postChannelName, // name
		"topic",         // type
		false,           // no durable
		false,           // auto-deleted
		false,           // internal
		false,           // no-wait
		nil,             // arguments
	)
	if err != nil {
		return err
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	routingKey := strconv.Itoa(int(userId))
	err = ch.PublishWithContext(ctx,
		postChannelName, // exchange
		routingKey,      // routing key
		false,           // mandatory
		false,           // immediate
		amqp.Publishing{
			ContentType: "text/plain",
			Body:        []byte(post),
		})

	if err != nil {
		return err
	}

	log.Printf(" [x] Sent %s", post)
	return nil
}

func (s *mqSenderImp) SendDialogMessage(ctx context.Context, data []byte) error {
	return s.sendDirectMessage(ctx, dialogWebsockQueueName, data)
}

func (s *mqSenderImp) SendUnreadDialogMessageIds(ctx context.Context, data []byte) error {
	return s.sendDirectMessage(ctx, counterQueueName, data)
}

func (s *mqSenderImp) SendUnreadDialogMessageIdsFailed(ctx context.Context, data []byte) error {
	return s.sendDirectMessage(ctx, counterDialogQueueName, data)
}

func (s *mqSenderImp) sendDirectMessage(_ context.Context, queueName string, data []byte) error {
	conn, err := amqp.Dial(os.Getenv("RABBITMQ_CONNECTION"))
	if err != nil {
		return err
	}
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil {
		return err
	}
	defer ch.Close()

	q, err := ch.QueueDeclare(
		queueName, // name
		false,     // durable
		false,     // delete when unused
		false,     // exclusive
		false,     // no-wait
		nil,       // arguments
	)
	if err != nil {
		return err
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	err = ch.PublishWithContext(ctx,
		"",     // exchange
		q.Name, // routing key
		false,  // mandatory
		false,  // immediate
		amqp.Publishing{
			ContentType: "text/plain",
			Body:        data,
		})
	if err != nil {
		return err
	}

	return nil
}
